src/coprocessor/dag/rpn_expr/types.rs

```diff
@@ -0,0 +1,695 @@
+            _ => None,
+        }
+    }
+}
+
+/// An RPN expression node list which represents an expression in Reverse Polish notation.
+#[derive(Debug)]
+pub struct RpnExpressionNodeVec(Vec<RpnExpressionNode>);
+
+impl std::ops::Deref for RpnExpressionNodeVec {
+    type Target = Vec<RpnExpressionNode>;
+
+    fn deref(&self) -> &Vec<RpnExpressionNode> {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for RpnExpressionNodeVec {
+    fn deref_mut(&mut self) -> &mut Vec<RpnExpressionNode> {
+        &mut self.0
+    }
+}
+
+impl RpnExpressionNodeVec {
+    /// Builds the RPN expression node list from an expression definition tree.
+    // TODO: Deprecate it in Coprocessor V2 DAG interface, since we should accept RPN list directly.
+    pub fn build_from_expr_tree(def: Expr, time_zone: Tz) -> Result<Self> {
+        let mut expr_nodes = Vec::new();
+        Self::append_rpn_nodes_recursively(
+            def,
+            time_zone,
+            &mut expr_nodes,
+            super::map_pb_sig_to_rpn_func,
+        )?;
+        Ok(RpnExpressionNodeVec(expr_nodes))
+    }
+
+    /// Transforms eval tree nodes into RPN nodes.
+    ///
+    /// Suppose that we have a function call:
+    ///
+    /// ```ignore
+    /// A(B, C(E, F, G), D)
+    /// ```
+    ///
+    /// The eval tree looks like:
+    ///
+    /// ```ignore
+    ///           +---+
+    ///           | A |
+    ///           +---+
+    ///             |
+    ///   +-------------------+
+    ///   |         |         |
+    /// +---+     +---+     +---+
+    /// | B |     | C |     | D |
+    /// +---+     +---+     +---+
+    ///             |
+    ///      +-------------+
+    ///      |      |      |
+    ///    +---+  +---+  +---+
+    ///    | E |  | F |  | G |
+    ///    +---+  +---+  +---+
+    /// ```
+    ///
+    /// We need to transform the tree into RPN nodes:
+    ///
+    /// ```ignore
+    /// B E F G C D A
+    /// ```
+    ///
+    /// The transform process is very much like a post-order traversal. This function does it
+    /// recursively.
+    fn append_rpn_nodes_recursively<F>(
+        mut def: Expr,
+        time_zone: Tz,
+        rpn_nodes: &mut Vec<RpnExpressionNode>,
+        fn_mapper: F,
+    ) -> Result<()>
+    where
+        F: Fn(tipb::expression::ScalarFuncSig) -> Result<Box<dyn RpnFunction>> + Copy,
+    {
+        // TODO: We should check whether node types match the function signature. Otherwise there
+        // will be panics when the expression is evaluated.
+
+        use crate::coprocessor::codec::mysql::{Decimal, Duration, Json, Time, MAX_FSP};
+        use crate::util::codec::number;
+        use std::convert::{TryFrom, TryInto};
+
+        let field_type = def.take_field_type();
+        let eval_type =
+            EvalType::try_from(field_type.tp()).map_err(|e| Error::Other(box_err!(e)))?;
+
+        match def.get_tp() {
+            ExprType::Null => {
```

Comment:
Could we create an inline function to create a nil `ScalarValue` with `EvalType`?

Reply:
Those functions are called only once, so that I don't think it's *very* necessary?